pkg/pulumiyaml/analyser.go

```diff
@@ -8,14 +8,255 @@ import (
+				from == schema.NumberType ||
+				from == schema.IntType)
+		case schema.AssetType:
+			// Some schema fields with given type Asset actually accept either
+			// Assets or Archives. We accept some invalid inputs instead of
+			// rejecting valid inputs.
+			return okIf(from == schema.AssetType || from == schema.ArchiveType)
+		default:
+			return okIf(from == to)
+		}
+	}
+
+	switch to := to.(type) {
+	case *schema.UnionType:
+		reasons := []*notAssignable{}
+		for _, subtype := range to.ElementTypes {
+			because := isAssignable(from, subtype)
+			if because == nil {
+				return nil
+			}
+			reasons = append(reasons, because)
+		}
+		return fail.Because(reasons...)
+	case *schema.ArrayType:
+		from, ok := from.(*schema.ArrayType)
+		if !ok {
+			return fail
+		}
+		return okIfAssignable(isAssignable(from.ElementType, to.ElementType))
+	case *schema.MapType:
+		switch from := from.(type) {
+		case *schema.MapType:
+			return okIfAssignable(isAssignable(from.ElementType, to.ElementType))
+		case *schema.ObjectType:
+			// YAML does not distinguish between maps and arrays, but our type system does.
+			// We allow implicit conversions from YAML objects into maps.
+			if len(from.Properties) == 0 {
+				// The object has no properties, so we coerce it's type to the type of the
+				// map trivially
+				return okIf(true)
+			}
+			for _, prop := range from.Properties {
+				notOk := isAssignable(prop.Type, to.ElementType)
+				if notOk != nil {
+					return fail.Because(notOk)
+				}
+			}
+			return okIf(true)
+		default:
+			return fail
+		}
+	case *schema.ResourceType:
+		from, ok := from.(*schema.ResourceType)
+		return okIf(ok && to.Token == from.Token)
+	case *schema.EnumType:
+		notAssignable := isAssignable(from, to.ElementType)
+		if notAssignable != nil {
+			return fail
+		}
+		// TODO: check that known enum values are type checked against valid
+		// values e.g. string "Foo" should not be assignable to
+		// type Enum { Type: string, Elements: ["fizz", "buzz"] }
+		return okIf(true)
+	case *schema.ObjectType:
+		// We implement structural typing for objects.
+		from, ok := from.(*schema.ObjectType)
+		if !ok {
+			return fail
+		}
+		for _, prop := range to.Properties {
+			fromProp, ok := from.Property(prop.Name)
+			if prop.IsRequired() && !ok {
+				return fail.Because(&notAssignable{reason: fmt.Sprintf("Missing required property '%s'", prop.Name)}).Property(prop.Name)
+			}
+			if !ok {
+				// We don't have an optional property, which is ok
+				continue
+			}
+			// We have a matching property, so the type must agree
+			notAssignable := isAssignable(fromProp.Type, prop.Type)
+			if notAssignable != nil {
+				return fail.Because(notAssignable).Property(prop.Name)
+			}
+		}
+		return nil
+	case *schema.TokenType:
+		if to.UnderlyingType != nil {
+			return isAssignable(from, to.UnderlyingType)
+		}
+		return &notAssignable{reason: fmt.Sprintf("Unknown opaque type: %s", to.Token), internal: true}
```

Comment:
These errors look good, but I'd like to see a test case for the diagnostics output.

Reply:
Check out https://github.com/pulumi/pulumi-yaml/pull/228/commits/2f7005f9ce139ae7933c4413974551c155d8bfc0 and https://github.com/pulumi/pulumi-yaml/pull/228/commits/5871d69f6821ec09b869a1a5476f68f0d6ebdc4f.